File: Algorithms/sorting-algorithms/shell-sort.cpp
#include "iostream"
using namespace std;

void shellSort(int * array, int n){
    for ( int interval = n /2 ; interval > 0; interval /= 2){
        for (int i = interval; i < n; i++){
            int temp = array[i];
            int j;
            for (j = i; j >= interval && array[j - interval] > temp; j -= interval){
                array[j] = array[j - interval];

            }
            array[j] = temp;
        }
    }
}

void printArray(int * array, ssize_t size){
    for (int i = 0; i < size; i++){
        cout << array[i] << " ";
    }   
    cout << endl;
}


int main(){
    int data[] = {-2, 45, 0, 11, -9};
    int size = sizeof(data) / sizeof(* data);
    shellSort(data, size);
    printArray(data, size);
}

File: Algorithms/sorting-algorithms/shell-sort.py
""" shell sort python implementation: """

def shellSort(array, n):
    """ rearrange intervals at n/4, n/2, n/8 intervals: """
    interval = n // 2
    while interval > 0:
        for i in range(interval, n):
            temp = array[i]
            j = i

            while j >= interval and array[j - interval] > temp:
                array[j] = array[j - interval]
                j -= interval
            array[j] = temp
        interval //= 2

def printArray(array):
    for i in range(len(array)):
        print(array[i], end=" ")
    print()

if __name__ == "__main__":
    data = [-2, 45, 0, 11, -9]
    size = len(data)

    shellSort(data, size)
    printArray(data)
